### PR TITLE
Moving CSI Driver to GA.

### PR DESCRIPTION
**Pull request by huffmanca**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Graduates the CSI Driver to GA. This only touches the API, their tests, and the associated staging files.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/enhancements/issues/770

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
The storage.k8s.io/CSIDriver has moved to GA, and is now available for use.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/768
```

### CODE REVIEW

File: staging/src/k8s.io/api/storage/v1/types.go [hunk 1]
```diff
@@ -221,6 +221,133 @@ type VolumeError struct {
 // +genclient:nonNamespaced
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// CSIDriver captures information about a Container Storage Interface (CSI)
+// volume driver deployed on the cluster.
+// CSI drivers do not need to create the CSIDriver object directly. Instead they may use the
+// cluster-driver-registrar sidecar container. When deployed with a CSI driver it automatically
```
Comment (jsafrane) [inline staging/src/k8s.io/api/storage/v1/types.go]: We do not support cluster-driver-registrar any longer: https://kubernetes-csi.github.io/docs/cluster-driver-registrar.html

**Review by jsafrane — COMMENTED** [review]
* I think you should also flip the feature gate to GA and remove any code that flips the feature gate to false (basically only in unit tests).
* A/D controller and kubelet should use storage/v1 CSIDriver instead of v1beta1. Basically nothing in kubelet and kube-controller-manager should import `k8s.io/api/storage/v1beta1` (at least not for CSIDriver).

https://github.com/kubernetes/kubernetes/pull/83474 is a good example to follow.

**Comment by jsafrane** [comment]
> *I think you should also flip the feature gate to GA and remove any code that flips the feature gate to false (basically only in unit tests).

\+ Jordan says "drop all associated checks of ...FeatureGate.Enabled(features.CSIDriver) and assume they evaluate to true now" https://github.com/kubernetes/kubernetes/pull/83568/files#r344719331

File: pkg/volume/csi/csi_plugin.go [hunk 2]
```diff
@@ -807,6 +763,7 @@ func (p *csiPlugin) supportsVolumeLifecycleMode(driver string, volumeMode storag
 		return fmt.Errorf("volume mode %q not supported by driver %s (only supports %q)", volumeMode, driver, csiDriver.Spec.VolumeLifecycleModes)
 	}
 }
+*/
```
Comment (bertinatto) [inline pkg/volume/csi/csi_plugin.go]: Uncomment?

File: pkg/features/kube_features.go [hunk 3]
```diff
@@ -561,7 +561,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	TopologyManager:                {Default: false, PreRelease: featuregate.Alpha},
 	ServiceNodeExclusion:           {Default: false, PreRelease: featuregate.Alpha},
 	NodeDisruptionExclusion:        {Default: false, PreRelease: featuregate.Alpha},
-	CSIDriverRegistry:              {Default: true, PreRelease: featuregate.Beta},
+	CSIDriverRegistry:              {Default: true, PreRelease: featuregate.GA},
```
Comment (bertinatto) [inline pkg/features/kube_features.go]: LockToDefault: true

**Comment by smarterclayton** [comment]
/assign @smarterclayton 

for api review

File: pkg/registry/storage/rest/storage_storage.go [hunk 4]
```diff
@@ -144,6 +144,15 @@ func (p RESTStorageProvider) v1Storage(apiResourceConfigSource serverstorage.API
 		storage["csinodes"] = csiNodeStorage.CSINode
 	}
 
+	// register csidrivers if CSIDriverRegistry feature gate is enabled
+	if utilfeature.DefaultFeatureGate.Enabled(features.CSIDriverRegistry) {
```
Comment (jsafrane) [inline pkg/registry/storage/rest/storage_storage.go]: The feature is enabled and can't be disabled, no need to check for it.

File: staging/src/k8s.io/api/storage/v1/types.go [hunk 5]
```diff
@@ -221,6 +221,133 @@ type VolumeError struct {
...
+	// PV/PVC mechanism.
+	// The other mode is "Ephemeral". In this mode, volumes are defined inline
+	// inside the pod spec with CSIVolumeSource and their lifecycle is tied to
+	// the lifecycle of that pod. A driver has to be aware of this
+	// because it is only going to get a NodePublishVolume call for such a volume.
+	// For more information about implementing this mode, see
+	// https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html
+	// A driver can support one or more of these modes and
+	// more modes may be added in the future.
+	// +optional
```
Comment (jsafrane) [inline staging/src/k8s.io/api/storage/v1/types.go]: Please add a note that this field is beta.

**Review by jsafrane — COMMENTED** [review]
All `features.CSIDriverRegistry` checks should be removed from the code, the feature should be always on
(see https://github.com/kubernetes/kubernetes/pull/84814/files#r385148181)

File: pkg/registry/storage/rest/storage_storage.go (hunk 4 above)
Comment (huffmanca) [reply pkg/registry/storage/rest/storage_storage.go]: I've removed these checks wherever they were found. I noticed that there are still some checks for other features (such as CSINodeInfo), should we create an issue to remove these checks as well?

File: pkg/features/kube_features.go [hunk 6]
```diff
@@ -585,7 +585,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	TopologyManager:                {Default: false, PreRelease: featuregate.Alpha},
 	ServiceNodeExclusion:           {Default: false, PreRelease: featuregate.Alpha},
 	NodeDisruptionExclusion:        {Default: false, PreRelease: featuregate.Alpha},
-	CSIDriverRegistry:              {Default: true, PreRelease: featuregate.GA},
+	CSIDriverRegistry:              {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
```
Comment (msau42) [inline pkg/features/kube_features.go]: Add a comment to remove in 1.20

File: test/integration/etcd/data.go [hunk 7]
```diff
@@ -503,13 +503,17 @@ func GetEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionRes
...
-		}
+	etcdStorageData[gvr("storage.k8s.io", "v1beta1", "csidrivers")] = StorageData{
+		Stub:             `{"metadata": {"name": "csid1"}, "spec": {"attachRequired": true, "podInfoOnMount": true}}`,
+		ExpectedEtcdPath: "/registry/csidrivers/csid1",
+	}
+
+	// k8s.io/kubernetes/pkg/apis/storage/v1
+	etcdStorageData[gvr("storage.k8s.io", "v1", "csidrivers")] = StorageData{
+		Stub:             `{"metadata": {"name": "csid2"}, "spec": {"attachRequired": true, "podInfoOnMount": true}}`,
+		ExpectedEtcdPath: "/registry/csidrivers/csid2",
```
Comment (msau42) [inline test/integration/etcd/data.go]: We need to keep the default GVK to be v1beta for 1 release to handle HA master version skew.

See [here](https://github.com/kubernetes/kubernetes/pull/83474/commits/603a2aa8a99869dc51ea1be2ce6e6ca10c23e55f#diff-a6932c2bd6398cd55df3be4564939a5eR559) and [here](https://github.com/kubernetes/kubernetes/pull/83474/commits/603a2aa8a99869dc51ea1be2ce6e6ca10c23e55f#diff-c74f8b85666a09e8b8a2a052f4a91abdR57)

Comment (msau42) [reply test/integration/etcd/data.go]: Oops, I see it's added in the 3rd commit. Would still be good to add a TODO and open an issue to track removing it next release.

File: pkg/features/kube_features.go (hunk 6 above)
Comment (huffmanca) [reply pkg/features/kube_features.go]: Comment has been included.

File: test/integration/etcd/data.go (hunk 7 above)
Comment (huffmanca) [reply test/integration/etcd/data.go]: I've updated this to include a TODO, and created issue https://github.com/kubernetes/kubernetes/issues/88821 to track it.

**Comment by jsafrane** [comment]
/lgtm
/approve
pkg/volume and other storage parts (i.e. not API)

**Comment by jsafrane** [comment]
snapshots tests are flaky and unrelated to this PR (https://github.com/kubernetes/kubernetes/pull/88801)

**Comment by msau42** [comment]
Minor nit on release note: Clarify it's the storage.k8s.io/CSIDriver object. It currently reads as if it's talking about some vendor csi driver.

**Comment by huffmanca** [comment]
> Minor nit on release note: Clarify it's the storage.k8s.io/CSIDriver object. It currently reads as if it's talking about some vendor csi driver.

The release note has been updated.

File: staging/src/k8s.io/api/storage/v1/types.go [hunk 8]
```diff
@@ -221,6 +221,133 @@ type VolumeError struct {
...
+	// "csi.storage.k8s.io/ephemeral" is a new feature in Kubernetes 1.16. It is only
+	// required for drivers which support both the "Persistent" and "Ephemeral" VolumeLifecycleMode.
+	// Other drivers can leave pod info disabled and/or ignore this field.
+	// As Kubernetes 1.15 doesn't support this field, drivers can only support one mode when
+	// deployed on such a cluster and the deployment determines which mode that is, for example
+	// via a command line parameter of the driver.
+	// +optional
+	PodInfoOnMount *bool `json:"podInfoOnMount,omitempty" protobuf:"bytes,2,opt,name=podInfoOnMount"`
+
+	// VolumeLifecycleModes defines what kind of volumes this CSI volume driver supports.
```
Comment (smarterclayton) [inline staging/src/k8s.io/api/storage/v1/types.go]: This should be `volumeLifecycleModes` - always reference the external name in external API doc.

File: staging/src/k8s.io/api/storage/v1/types.go [hunk 9]
```diff
@@ -221,6 +221,133 @@ type VolumeError struct {
...
+	// PV/PVC mechanism.
+	// The other mode is "Ephemeral". In this mode, volumes are defined inline
+	// inside the pod spec with CSIVolumeSource and their lifecycle is tied to
+	// the lifecycle of that pod. A driver has to be aware of this
+	// because it is only going to get a NodePublishVolume call for such a volume.
+	// For more information about implementing this mode, see
+	// https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html
+	// A driver can support one or more of these modes and
+	// more modes may be added in the future.
+	// This field is beta.
```
Comment (smarterclayton) [inline staging/src/k8s.io/api/storage/v1/types.go]: This is not actually true after this PR?  Or is it?

Comment (smarterclayton) [reply staging/src/k8s.io/api/storage/v1/types.go]: Isn't there another PR in flight that changes the description of this field?  If so, you need to incorporate those changes before I'll approve it (I *think* it got approved yesterday).

**Review by smarterclayton — CHANGES_REQUESTED** [review]
Some final things that need to be cleared up

File: staging/src/k8s.io/api/storage/v1/types.go (hunk 9 above)
Comment (huffmanca) [reply staging/src/k8s.io/api/storage/v1/types.go]: The `volumeLifecycleModes` field is still beta, while the rest of the types should be GA

Comment (gnufied) [reply staging/src/k8s.io/api/storage/v1/types.go]: Did a PR change description of `volumeLifecycle` mode recently? I might have missed it. @smarterclayton which PR you are referring to? cc @msau42

File: staging/src/k8s.io/api/storage/v1/types.go (hunk 8 above)
Comment (huffmanca) [reply staging/src/k8s.io/api/storage/v1/types.go]: This has been updated with the latest push.

File: staging/src/k8s.io/api/storage/v1/types.go (hunk 9 above)
Comment (huffmanca) [reply staging/src/k8s.io/api/storage/v1/types.go]: I looked and wasn't able to find a PR that changed this description, either open or recently merged.

**Comment by huffmanca** [comment]
@smarterclayton ,

After discussing this with @gnufied  further, I've removed `listType` from the Items[] field and added a single API violation for this type.

**Comment by smarterclayton** [comment]
/approve

API change is approved.

File: staging/src/k8s.io/api/storage/v1/types.go [hunk 10]
```diff
@@ -221,6 +221,131 @@ type VolumeError struct {
...
+	// defined by the CSI specification and implemented in Kubernetes via the usual
+	// PV/PVC mechanism.
+	// The other mode is "Ephemeral". In this mode, volumes are defined inline
+	// inside the pod spec with CSIVolumeSource and their lifecycle is tied to
+	// the lifecycle of that pod. A driver has to be aware of this
+	// because it is only going to get a NodePublishVolume call for such a volume.
+	// For more information about implementing this mode, see
+	// https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html
+	// A driver can support one or more of these modes and
+	// more modes may be added in the future.
```
Comment (msau42) [inline staging/src/k8s.io/api/storage/v1/types.go]: I don't see the note that this field is beta?

Comment (gnufied) [reply staging/src/k8s.io/api/storage/v1/types.go]: It is there below your comment.

Comment (msau42) [reply staging/src/k8s.io/api/storage/v1/types.go]: i swear it wasn't there..... ok :-)

**Comment by irvifa** [comment]
@huffmanca I think you need to rebase this PR first.. There's a conflict on the `csi_mounter`

**Pull request merged**